src/MongoDB.Driver/MongoCollectionImpl.cs

```diff
@@ -155,6 +155,50 @@ public override async Task<IAsyncCursor<TResult>> AggregateAsync<TResult>(IClien
             }
         }
 
+        public override void AggregateToCollection<TResult>(PipelineDefinition<TDocument, TResult> pipeline, AggregateOptions options, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            UsingImplicitSession(session => AggregateToCollection(session, pipeline, options, cancellationToken), cancellationToken);
+        }
+
+        public override void AggregateToCollection<TResult>(IClientSessionHandle session, PipelineDefinition<TDocument, TResult> pipeline, AggregateOptions options, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            Ensure.IsNotNull(session, nameof(session));
+            var renderedPipeline = Ensure.IsNotNull(pipeline, nameof(pipeline)).Render(_documentSerializer, _settings.SerializerRegistry);
+            options = options ?? new AggregateOptions();
+
+            var lastStage = renderedPipeline.Documents.LastOrDefault();
+            var lastStageName = lastStage?.GetElement(0).Name;
+            if (lastStage == null || (lastStageName != "$out" && lastStageName != "$merge"))
+            {
+                throw new InvalidOperationException("AggregateToCollection requires that the last stage be $out or $merge.");
+            }
+
+            var aggregateOperation = CreateAggregateToCollectionOperation(renderedPipeline, options);
+            ExecuteWriteOperation(session, aggregateOperation, cancellationToken);
+        }
+
+        public override Task AggregateToCollectionAsync<TResult>(PipelineDefinition<TDocument, TResult> pipeline, AggregateOptions options, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return UsingImplicitSessionAsync(session => AggregateToCollectionAsync(session, pipeline, options, cancellationToken), cancellationToken);
+        }
+
+        public override async Task AggregateToCollectionAsync<TResult>(IClientSessionHandle session, PipelineDefinition<TDocument, TResult> pipeline, AggregateOptions options, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            Ensure.IsNotNull(session, nameof(session));
+            var renderedPipeline = Ensure.IsNotNull(pipeline, nameof(pipeline)).Render(_documentSerializer, _settings.SerializerRegistry);
+            options = options ?? new AggregateOptions();
+
+            var lastStage = renderedPipeline.Documents.LastOrDefault();
+            var lastStageName = lastStage?.GetElement(0).Name;
+            if (lastStage == null || (lastStageName != "$out" && lastStageName != "$merge"))
+            {
+                throw new InvalidOperationException("AggregateToCollectionAsync requires that the last stage be $out or $merge.");
+            }
+
+            var aggregateOperation = CreateAggregateToCollectionOperation(renderedPipeline, options);
+            await ExecuteWriteOperationAsync(session, aggregateOperation, cancellationToken).ConfigureAwait(false);
```

Comment:
I guess it's in line with the rest of the methods in this class, but can you explain why lines 187 to 198 aren't factored into a common method to share with the sync version?  They seem identical.

Reply:
Fair question.

I just copy pasted from the existing method.

Do you want me to refactor? And if so, should I refactor the other methods also?

Reply:
No, I just wanted to know if there was a reason. Let's leave it alone for now.

Reply:
Agreed with Jeff. It sounds like the only difference is whether we need to call operation in sync or async way